Merge THM Module into MOOSE

permcody opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->

Merging the THM repository into the MOOSE repository. It has been approved for OSS release

## Design
<!--A concise description (design) of the enhancement.-->

See the THM design documents and SDR.

## Impact
<!--Will the enhancement change existing APIs or add something new?-->

This PR adds a completely new physics module (Major)

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

andrsd committed:
Updating heat_transfer_from_external_app_1phase tests

andrsd committed:
Updating inlet_density_velocity_1phase tests

andrsd committed:
Updating inlet_mass_flow_rate_1phase tests

andrsd committed:
Converting Pump1Phase into AD

andrsd committed:
Updating vector_velocity test

andrsd committed:
Converting BoundaryFlux3EqnFunction into AD

andrsd committed:
Updating tests

andrsd committed:
DirectionMaterial does not have to be AD

andrsd committed:
D_h is regular material property

andrsd committed:
Converting FanningFrictionFactorMaterial into AD

Refs #19868

andrsd committed:
AD version of Dean number

Refs #19868

andrsd committed:
Fixing Grashof function signature

andrsd committed:
Updating test actions to use AD

Refs #19868

andrsd committed:
Heat flux setup moved in addMooseObjects

Refs #19868

andrsd committed:
Allow non-AD version of Shaft and ShaftConnectedMotor

andrsd committed:
Merge branch 'andrsd/ad' into 'devel'

Conversion to AD

Closes #19868
See merge request idaholab/thm!563

andrsd committed:
Fixing heavy tests

refs #0

aeslaughter committed:
Remove duplicate ids

(refs #58)

andrsd committed:
Merge branch 'docs-58' into 'devel'

Remove duplicate ids

See merge request idaholab/thm!564

Commit:
Submodule update

Refs #0

licharlot committed:
Modified residuals for loss coefficient in volume junction

- Added term on energy equation
- Use the conditions from the first connected pipe instead of the junction conditions
- A_ref used only if provided
- Updated and regolded  the tests

Refs #209, #19890

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!565

Commit:
Merge commit '19ac1d8d7dda402757d42e18812d4fa4a697bab6'

andrsd committed:
Adding migration guide for AD

Refs #19868

andrsd committed:
AD version of ConvectiveHeatFlux1PhaseAux

Refs #0

andrsd committed:
Merge branch 'andrsd/sabertooth-fix' into 'devel'

AD version of ConvectiveHeatFlux1PhaseAux

See merge request idaholab/thm!567

Commit:
Merge commit '4c31a713781e32e3d3aead3c6a458f67f447535e'

andrsd committed:
Adding mailmap file

refs #0

andrsd committed:
Do not set processor ID in elements

Partitioner will assign the right processor ID.

Refs #19899

andrsd committed:
Only output errors on CPU 0

Refs #19899

andrsd committed:
Remove parallel restriction from error checking tests

Refs #19899

andrsd committed:
Updating features list on index page

Refs #0

andrsd committed:
Merge branch 'andrsd/thm-ad-guide' into 'devel'

Adding migration guide for AD

See merge request idaholab/thm!566

Commit:
Merge commit '7c2ad58d574f15153821c53d99b062b2bc8fb197'

licharlot committed:
Removed redundant tests

The component and user object associated with JunctionParallelChannels1Phase inherit from VolumeJunction1Phase.
The form losses are already tested in the VolumeJunction1Phase tests

Refs #0

licharlot committed:
Increased absolute 0 to 1e-8 for a test

Refs #0

licharlot committed:
Added documentation for form losses in VolumeJunction1phase

Closes #209, #19890

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!570

Commit:
Merge commit '17ab167aab5015ae0c1dfc52d49820827d8e1347'

andrsd committed:
Merge branch 'charlc/K' into 'devel'

Modified residuals for loss coefficient in volume junction

Closes #19890and #19889
See merge request idaholab/thm!569

Commit:
Merge commit 'cee3de83f443a1e5cdc8f0ed1e3b87ded17c1105'

andrsd committed:
Merge branch 'andrsd/mpi-log-fix' into devel

Commit:
Merge commit 'b5e78adb956ab1d1f2e10378ee153c4d748b475d'

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!571

Commit:
Merge commit 'd8ff271c98e0c0dec6619948eedf6de24b3cdb02'

andrsd committed:
Making control logic values restartable

If simulation crashes, we recover with the correct state of the control
system.

Closes #19873

joshuahansel committed:
Implemented AD for slope reconstruction

Closes #19782

joshuahansel committed:
Used solve_type=NEWTON in implicit Sod shock tube test

Refs #19782

andrsd committed:
Merge branch 'hansje/ad-slopes' into 'devel'

Implemented AD for slope reconstruction

Closes #19782
See merge request idaholab/thm!573

Commit:
Merge commit '4ded62f26445b1a78b507beff8bb79e4ad96c355'

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!574

Commit:
Merge commit 'b4d4d1cfd6f5a5732f346ba3f88fb689e98a59df'

andrsd committed:
Removing distributed mesh restriction from tests

Seems that the !570 fixed the problem with 3D HS on distributed mesh

Closes #19862

andrsd committed:
Merge branch 'andrsd/184' into 'devel'

Removing distributed mesh restriction from tests

Closes #19862
See merge request idaholab/thm!575

Commit:
Merge commit 'f07eb57e7f528ef62dc0e871ae5749406df4d4ab'

andrsd committed:
Guard against Re=0 in ADWallHeatTransferCoefficient3EqnDittusBoelterMaterial

When the code compute Re=0, a division by zero occurs inside the AD system.
Using limitation to prevent Reynolds number to be exactly zero. The minimum
value of 1. is being used to be consistent with the rest of the code (for
example TRACE closures use the same limitation)

Closes #19902

loganharbour committed:
Add coverage configuration file

refs #19903

andrsd committed:
Merge branch 'cov_config_223' into 'devel'

Add coverage configuration file

Closes #19903
See merge request idaholab/thm!577

andrsd committed:
Merge branch 'andrsd/222' into 'devel'

Guard against Re=0 in ADWallHeatTransferCoefficient3EqnDittusBoelterMaterial

Closes #19902
See merge request idaholab/thm!576

Commit:
Merge commit '3d39660b6e695344619cec8e34921f7b6b89f66a'

Commit:
Submodule update

Refs #0

Commit:
Merge commit 'c7b5ed3381b5bb33deefbfe0cacba44edb1ba8db'

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!578

Commit:
Merge commit '39bb26c8e09613c18c846b4e293a4f4a99e76eb7'

andrsd committed:
Fixing up PIDControl for recover/restart

Expressions like: `a = a + delta` need to be avoided in `execute`.
They lead to incorrect behavior during restart/recover.  The
correct form is `a = a_old + delta`.  This is consistent with
the other MOOSE systems like Postprocessors, etc.

Refs #19873

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!579

Commit:
Merge commit '5e75812b2006308b5b4db67b2458502d16a9163d'

andrsd committed:
Adding test

andrsd committed:
Fixing headings in CSS

- change size
- add space after sections

andrsd committed:
Adding CSS class for numbered steps

andrsd committed:
Adding CSS for warnings, tips and notes

andrsd committed:
Adding simple alert extension

andrsd committed:
Adding styles for footer

andrsd committed:
Using template for tutorials

andrsd committed:
Align the next button to content section

joshuahansel committed:
Documented relevant inlet components

Refs #19905

andrsd committed:
CSS classes for enumerate

Refs #19737

andrsd committed:
Tutorial: the basics

Refs #19737

joshuahansel committed:
Merge branch 'issue/195' into 'devel'

Making control logic values restartable

Closes #19873
See merge request idaholab/thm!572

Commit:
Merge commit 'f8370e715bd04bbe01c51c770d508b45cfa06752'

joshuahansel committed:
Added documentation on flow boundary connections

Refs #19905

andrsd committed:
Merge branch 'hansje/inlet-documentation' into 'devel'

Documented relevant inlet components

See merge request idaholab/thm!580

Commit:
Merge commit 'd065ea4d408cfd894fa47c87f45466454e0b2e78'

andrsd committed:
Merge branch 'andrsd/doco' into 'devel'

Doco improvements

See merge request idaholab/thm!581

Commit:
Merge commit '727e4a0e9f44a6c06ffd4db54ba43c6acaef0675'

andrsd committed:
Submodule update

Refs #0

Commit:
Merge commit 'c0b7b72f23062f559e7fefb9c16f723f8a6a0657'

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!583

Commit:
Merge commit 'f18587efb4c61718703ce4c48fe7cf3835581171'

licharlot committed:
Impose isentropic pressure drop in simple turbine model

Refs #19888

licharlot committed:
Modified SimpleTurbine1Phase to inherit from JunctionParallelChannels1Phase

Refs #19888

licharlot committed:
Updated simple turbine tests

- Changed conditions to make the calculation of the analytic pressure drop easier
- Added a check on the momentum change across the turbine
- Regolded as the solution has changed

Refs #19888

licharlot committed:
Added doco for SimpleTurbine1Phase

Ref #208, #19905

andrsd committed:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!585

Commit:
Merge commit '23ddcec24b41c57d7c2ca2b26e703f5b0fb9b83a'

joshuahansel committed:
Merge branch 'charlc/simple_turbine' into 'devel'

Impose isentropic pressure drop in simple turbine model

See merge request idaholab/thm!584

Commit:
Merge commit 'a6d97ef098491663a74d13ce91db66709020984c'

andrsd committed:
Merge branch 'andrsd/tut-basics' into 'devel'

Tutorial: the basics

See merge request idaholab/thm!582

andrsd committed:
Dark mode for doco

Documentation should be legible when users/developers have dark mode on.

Refs #19737

Commit:
Merge commit '573e3f71e3a2a1c345a3bcbecf51b13a68e4a52a'

andrsd committed:
Merge branch 'andrsd/dark-mode' into 'devel'

Dark mode for doco

See merge request idaholab/thm!586

Commit:
Merge commit 'b5ee8d758c7881679455ccfb10d23faad21fca17'

joshuahansel committed:
Added HeatRateRadiation post-processor

Refs #19897

joshuahansel committed:
Added HeatRateConvection post-processor

Closes #19897

joshuahansel committed:
Inherited RZ versions of convection and radiation PPs from non-RZ

Refs #19897

Commit:
Submodule update

Refs #0

andrsd committed:
Removing unused parameter

This is a left over from the task of removing CG.

Closes #19904

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!588

Commit:
Merge commit '4d2405a99d8b39aec66c286bc0e33d78bc7d7de5'

andrsd committed:
Merge branch 'andrsd/224' into 'devel'

Removing unused parameter

Closes #19904
See merge request idaholab/thm!589

Commit:
Merge commit '1407fa4dc8421d2299675337b30ffdfc60fd5d96'

andrsd committed:
Adding input files

andrsd committed:
Merge branch 'hansje/heat-rate-radiation' into 'devel'

Added HeatRateRadiation and HeatRateConvection post-processor

Closes #19897
See merge request idaholab/thm!587

Commit:
Merge commit '816ebffdd1e49058f5ae3dac17ea82fab4caf48c'

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!590

Commit:
Merge commit '4703c8dede33905f95ccc8eb62108220d52a00a5'

andrsd committed:
Updating README.md

Refs #0

Commit:
Merge commit 'c6bbdb0bc39080b5556d9fe3f60930253e5f0917'

andrsd committed:
Adding mrproper target to really clean the dir

Refs #0

Commit:
Merge commit '00a92bebf1173f08fca4074c2e8d2a9baa9731e4'

andrsd committed:
Tutorial text

Refs #19737

andrsd committed:
Renaming Enum.h to THMEnums.h

Make this file name unique, so that the build system always pick the right one.

Refs #0

andrsd committed:
Merge branch 'andrsd/include-fix' into 'devel'

Submodule update

See merge request idaholab/thm!592

andrsd committed:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!593

Commit:
Merge commit '4a25457f7801a667ca6c78dce641365c0a07b0f5'

andrsd committed:
Respect the -r command line options

If -r is specified we need to uniformly refine the mesh.

Closes #19906

andrsd committed:
Merge branch 'andrsd/226' into 'devel'

Respect the -r command line options

Closes #19906
See merge request idaholab/thm!594

Commit:
Merge commit '47fbf9372d34808168fdd14dccc1083bfff816e0'

joshuahansel committed:
Documented HeatStructureCylindrical

Refs #19905

joshuahansel committed:
Used Elem::vertex_average() instead of Elem::centroid()

Elem::centroid() is deprecated.

Refs #0

licharlot committed:
Adding a transition for the sign of the frictional torque when the shaft speed nears 0.

Closes #19907

andrsd committed:
Merge branch 'hansje/doc-hs' into 'devel'

Documented HeatStructureCylindrical

See merge request idaholab/thm!596

andrsd committed:
Merge branch 'hansje/update-centroid' into 'devel'

Used Elem::vertex_average() instead of Elem::centroid()

See merge request idaholab/thm!597

Commit:
Merge commit 'e761ed4efd7349c48a2c92b6f250e874389d9af7'

Commit:
Merge commit '20f69902832a26ba763895f72185964b012b8575'

andrsd committed:
Merge branch 'charlc/coastdown' into 'devel'

Adding a transition for the sign of the frictional torque when the shaft speed nears 0.

Closes #19907
See merge request idaholab/thm!598

Commit:
Submodule update

Refs #0

Commit:
Merge commit 'b21a6a8cd3dcac99545c31b426f37d0eb54e8388'

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!599

Commit:
Merge commit '225ddeb2981863415cc6542a8fe2b196c173999d'

joshuahansel committed:
Moved 2D heat structure documentation to common file

This common file will be used by HeatStructurePlate

Refs #19905

joshuahansel committed:
Documented HeatStructurePlate

Refs #19905

joshuahansel committed:
Documented HSBoundarySpecifiedTemperature

Refs #19905

joshuahansel committed:
Documented HSBoundaryAmbientConvection

Refs #19905

joshuahansel committed:
Documented HSBoundaryHeatFlux

Refs #19905

joshuahansel committed:
Documented HSBoundaryRadiation

Refs #19905

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!601

andrsd committed:
Merge branch 'hansje/more-hs-doc' into 'devel'

Documented more heat structure components

See merge request idaholab/thm!600

Commit:
Merge commit '31b8d442acec23beda1d7d7fd56bd1c533e9e1c4'

Commit:
Merge commit '4a919ee94769e3aa85203dc68f4e53f516ac2115'

andrsd committed:
Adding placeholders for some pages

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!602

Commit:
Merge commit 'a30e2fbdab813f0043a158fcd3d1f0ce95653823'

loganharbour committed:
Remove legacy input parameter construction

refs #19910

andrsd committed:
Merge branch 'harblh_param' into 'devel'

Remove legacy input parameter construction

Closes #19910
See merge request idaholab/thm!603

Commit:
Merge commit 'baf51276f9a56b5b12f9074f12a70d173bee9dce'

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!604

Commit:
Merge commit '578e66492333ffac8d891fbe004132939ebc08e0'

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!605

Commit:
Merge commit '10d9956d80781afa69c1c254f4a5a1dde59dcd5b'

andrsd committed:
Merge branch 'andrsd/spf-tut' into 'devel'

Single phase flow tutorial

See merge request idaholab/thm!591

Commit:
Merge commit '22d083df018ff139e2fa9e15411b33347ddaf29c'

joshuahansel committed:
Added API to Simulation for adding, getting, and checking for closures

Refs #19682

joshuahansel committed:
Added action to add closures

Refs #19682

joshuahansel committed:
Added ability to use closures created in input file to flow channels

The ability to use enumerated closures options has been preserved.
If the provided closures parameter corresponds to closures created
via the input file, then those closures are used. Else the enumerated
options are tried; if it corresponds to an enumerated option, that is
used, and a deprecated message is given. If the closures parameter
does not correspond to either method, then an error occurs.

Closes #19682

joshuahansel committed:
Moved closures to their own directory

Refs #19682

joshuahansel committed:
Added tests for using closure enumeration options

Refs #19682

joshuahansel committed:
Switched input files to use new closures syntax

Refs #19682

joshuahansel committed:
Added documentation for closures classes

Refs #19682

joshuahansel committed:
Updated tutorials with new closures setup

Refs #19682

joshuahansel committed:
Added news entry for new closures setup

Refs #19682

joshuahansel committed:
Used TimeIntegratedPostprocessor instead of TotalVariableValue

TotalVariableValue was deprecated in favor of TimeIntegratedPostprocessor

Refs #0

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!607

Commit:
Merge commit '277bbf00b79e2c406e0d37156e90774492c45caa'

joshuahansel committed:
Added method to FlowChannelBase to get connected heat transfer names

Closes #19914

joshuahansel committed:
Merge branch 'hansje/get-ht-names' into 'devel'

Added method to FlowChannelBase to get connected heat transfer names

Closes #19914
See merge request idaholab/thm!608

Commit:
Merge commit '67e9ceef701226a5034ed721652ae387a81bce5d'

joshuahansel committed:
Merge branch 'hansje/closures-block' into 'devel'

Created closures in the input file

Closes #19682
See merge request idaholab/thm!606

Commit:
Merge commit 'c41a467b5d39e0cd358117cc88a1fd5cbe0241f0'

andrsd committed:
Fixing a typo in single-phase flow tutorial

Quantity produced by PIDController is `pid:output`

Refs relap-7#758

andrsd committed:
Fixing type of jct6

jct6 is a in a bend so JunctionParallelChannels1Phase is just plain wrong.
I blame David for this.

Refs relap-7#758

andrsd committed:
Adding a note about fluid mixtures

Refs relap-7#758

joshuahansel committed:
Changed closures API to avoid overloading virtuals

Closes #19915

joshuahansel committed:
Merge branch 'hansje/overloaded-virtual' into 'devel'

Changed closures API to avoid overloading virtuals

Closes #19915
See merge request idaholab/thm!609

Commit:
Merge commit '5f3e145114e93890b7d66393ecbffaae71617691'

andrsd committed:
Merge branch 'tut-fixups' into 'devel'

Single phase flow tutorial fixes

See merge request idaholab/thm!610

Commit:
Merge commit '832672b3e3c679ea155ec4e25e3ff39dbd7990e4'

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!611

Commit:
Merge commit 'e49920d8c7a75c88c93fa57626f29effe1d41301'

permcody committed:
Modify Makefile to include thm module

Commit:
Submodule update

Refs #0

andrsd committed:
Merge branch 'submodule-update' into 'devel'

Submodule update

See merge request idaholab/thm!612

Commit:
Merge commit '12d92f6ae72159d2f01196930c6cc054fb00fe19'

permcody committed:
Source changes to add THM module to MOOSE modules

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/922067/) on 10666af wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19916/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19916/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1c4a0ba58e8a10e4900a0bc6dbea0ea973530207`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/922067/) on 10666af wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

permcody committed:
Clang Format Fixes

refs #19916

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/922104/) on 89ed13b wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

loganharbour commented:
- Remove `.coverage` and add to the root `.coverage` file with `[thm]` instead of `[default`]
- Remove .gitignore

loganharbour commented:
Script used for porting issues in case we need it later:

```python
import gitlab
from github import Github
import time
from datetime import datetime, timedelta
import re
import csv

gitlab_url = 'https://gitlab.software.inl.gov'
gitlab_token = ''
github_token = ''
repo = 'idaholab/thm'
to_repo = 'idaholab/moose'

gl = gitlab.Gitlab(gitlab_url, private_token=gitlab_token)
gh = Github(github_token)

gl_repo = gl.projects.get(repo)
gh_repo = gh.get_repo(to_repo)

# To be filled with old issue ID -> new issue ID
issue_map = {}

# Name map for the header from person -> github persona
author_map = {'Alexander Lindsay': 'lindsayad',
            'David Andrs': 'andrsd',
            'Fande Kong': 'fdkong',
            'Guillaume Giudicelli': 'GiudGiud',
            'Joshua Hansel': 'joshuahansel',
            'Logan Harbour': 'loganharbour',
            'Sebastian Schunert': 'snschune'}

# User map from gitlab -> github persona
user_map = {'hansje': 'joshuahansel',
           'balep': 'pbalest',
           'schuseba': 'snschune',
           'schus': 'snschune',
           'marcm': 'rcm59',
           'KONGF': 'fdkong',
           'lindad': 'lindsayad',
           'ortej': 'jortensi',
           'LABOVM': 'vincentlaboure',
           'rcm': 'rcm59',
           'slauae': 'aeslaughter'}

# Converts a gitlab time -> a human time (this is dirty but it works...)
def human_time(timestamp):
    created = datetime.strptime(timestamp[:-5], '%Y-%m-%dT%H:%M:%S') - timedelta(hours=7)
    return created.strftime('%b %d, %Y %I:%M%p MST')

# Replaces an author with a name to their github username
def replace_author(object):
    if object.author['name'] in author_map:
        return '@' + author_map[object.author['name']]
    return object.author['name']

# Replaces @ mentions
def replace_mentions(str):
    for from, to in user_map.items():
        str = str.replace('@' + from, '@' + to)
    return str

# Header for adding a comment for an issue/note/etc
def body_header(issue, object, extra_content=None):
    new_body = 'Imported from [{}#{}]({}) '.format(repo, issue.iid, issue.web_url)
    new_body += 'by {} at {}\n\n'.format(replace_author(object), human_time(object.created_at))
    if extra_content:
        new_body += extra_content + '\n\n'
    new_body += '------\n\n'
    return new_body

# Gets a URL to a merge request on a gitlab instance
def gitlab_merge_request_url(num):
    return gitlab_url + '/' + repo + '/-/merge_requests/{}'.format(num)

# Helper for calling a github action that can be rate limited; will try it every 30 sec
# until the rate liminting is gone
def do_github(action):
    while True:
        try:
            result = action()
            time.sleep(30)
            return result
        except:
            print('Rate limited')
            time.sleep(30)

# Issues that have uploads that we need to manually change
has_uploads = set()

issues = sorted(gl_repo.issues.list(all=True), key=lambda entry: entry.iid)
print('Importing {} issues...'.format(len(issues)))
for i in range(len(issues)):
    issue = issues[i]
    print('{}/{}: !{} "{}"'.format(i + 1, len(issues), issue.iid, issue.title))

    # Whether or not the issue was closed by a merge request
    # (to trigger a MR closed comment instead of a generic comment)
    closed_by_merge_request = None
    # The merge requests that were mentioned in the issue
    mentioned_merge_requests = set()
    # Comments we need to add
    comments = []

    # Work through all of the notes, which could be mentions or actual user comments
    for note in sorted(issue.notes.list(all=True), key=lambda entry: datetime.strptime(entry.created_at[:-5], '%Y-%m-%dT%H:%M:%S').timestamp()):
        # System notes are mentions (MRs, commits, etc)
        if note.system:
            search_closed_merge_request = re.search('closed via merge request !([0-9]+)', note.body)
            if search_closed_merge_request:
                mr_num = search_closed_merge_request.group(1)
                closed_by_merge_request = mr_num
                mentioned_merge_requests.add(mr_num)

            search_mentioned_merge_request = re.search('mentioned in merge request !([0-9]+)', note.body)
            if search_mentioned_merge_request:
                mr_num = search_mentioned_merge_request.group(1)
                mentioned_merge_requests.add(mr_num)
        # Standard comment by a user
        else:
            # Keep track of issues where we need to manually add uploads
            if '/uploads' in note.body:
                has_uploads.add(issue.iid)

            comment_body = body_header(issue, note) + replace_mentions(note.body)
            comments.append(comment_body)

    # If not closed by a MR but it was closed, add a comment as to when it was closed
    if issue.state == 'closed' and closed_by_merge_request is None:
        comments.append('Closed at {}.'.format(human_time(issue.closed_at)))

    # Create the extra content in the header for mentioned MRs
    extra_content = None
    if mentioned_merge_requests:
        extra_content = 'Mentioned merge requests:\n\n'
        for mr_num in sorted(mentioned_merge_requests):
            mr = gl_repo.mergerequests.get(mr_num)
            link = '[{} (!{})]({})'.format(mr.title, mr_num, gitlab_merge_request_url(mr_num))
            extra_content += '- ' + link + '\n'
            if mr_num == closed_by_merge_request:
                comments.append('Closed by merge request ' + link + ' at ' + human_time(mr.merged_at))

    # Keep track of issues where we need to manually add uploads
    if '/uploads' in issue.description:
        has_uploads.add(issue.iid)

    # Create the issue
    issue_body = body_header(issue, issue, extra_content) + replace_mentions(issue.description)
    create_issue = lambda: gh_repo.create_issue(title='THM: ' + issue.title, body=issue_body, labels=['C: Modules'])
    new_issue = do_github(create_issue)

    # Save the new issue number to the old -> new ID map
    with open('issue_map.csv', 'a') as f:
        csv.writer(f).writerow((issue.iid, new_issue.number))

    # Post all of the comments
    for comment in comments:
        do_github(lambda: new_issue.create_comment(comment))

    # Close the issue if it's closed
    if issue.state == 'closed':
        do_github(lambda: new_issue.edit(state='closed'))

if has_uploads:
    print('\nThe following issue(s) have uploads and will need to be edited manually:')
    for issue_id in sorted(has_uploads):
        issue = gl_repo.issues.get(issue_id)
        print('- !{} "{}"'.format(issue_id, issue.title))
```

permcody committed:
Removing old module cruft

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/924060/) on 66e3aa4 wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/924134/) on 7233daa wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

permcody committed:
Merge remote-tracking branch 'remotes/thm_source/master' into thm_merge_fixes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/924243/) on 3ce9603 wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

loganharbour commented on modules/modules.mk:
```diff
@@ -31,55 +31,60 @@ ifeq ($(ALL_MODULES),yes)
+        RDG                         := yes
 endif
 
-# Modules that follow have one or more dependencies
```
Why remove this comment? Breaking the point where the includes aren't independent anymore I think is good

loganharbour commented on modules/thm/include/base/THMApp.h:
```diff
@@ -65,6 +65,13 @@ class THMApp : public MooseApp
   static void registerApps();
   static void registerAll(Factory & f, ActionFactory & af, Syntax & s);
 
+  /**
+   * Deprecated Methods
+   */
+  static void registerObjects(Factory & factory);
+  static void associateSyntax(Syntax & syntax, ActionFactory & action_factory);
+  static void registerExecFlags(Factory & factory);
+
```
I say don't even include these if we didn't have them in THM before

loganharbour reviewed (changes requested):
Remove

- `modules/thm/.clang-format`
- `modules/thm/.coverage`
- `modules/thm/.gitlab`
- `modules/thm/.mailmap`

permcody committed:
Missed the Combined App changes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/924500/) on 3714d4a wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

permcody committed:
Another round of refactoring

The Test and 'regular' app aren't registering things correctly.
The App should do all of the registration including all dependencies (it was not).
The TestApp should regsiter the test objects and the main app. Still seeing
failures with combined on my local system.

refs #19916

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/924937/) on fd57bcc wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/924936/) on fd57bcc wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19916/site)

This comment will be updated on new commits.

permcody committed:
Cleanup of the CombinedTestApp

How does any of this work? We are missing a ton of module test app registrations. I guess nobody
builds test objects?

refs #19916

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/926072/) on 885865f wanted to post the following:

Warning: This PR changes repo size by 11.56 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/926106/) on c303301 wanted to post the following:

Warning: This PR changes repo size by 11.56 MiB.

permcody committed:
Manually registering THMApp as a buildable app in Combined

refs #19916

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/927217/) on 3c1e9ef wanted to post the following:

Warning: This PR changes repo size by 11.56 MiB.

permcody committed:
Fixing bad test spec

refs #19916

permcody committed:
Fixing up unit tests

refs #19916

lindsayad commented:
Wow @andrsd is going to build an insurmountable commit count lead in MOOSE once this is merged 😄 ~A question I would ask is whether it's universally known that "thermal hydraulics" means essentially system thermal hydraulics. I feel like the answer is probably yes, but if it wasn't then this could be confusing to users that we now have a thermal hydraulics module that is separate from things like `heat_conduction` (whose name should probably be updated to be more like `heat_transfer`) and `navier_stokes`.~ Belay my concerns. I see that THM already pulls in those aforementioned modules, so I think it's fair to consider this to be a cohesive (albeit mostly 1D) thermal hydraulics module.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/927770/) on 5beffb4 : invalidated by @permcody

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/927770/) on 5beffb4 wanted to post the following:

Warning: This PR changes repo size by 11.56 MiB.

permcody committed:
Restrict many tests to run only with global indexing

refs #19916

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/928073/) on 5199aeb wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

cticenhour reviewed (dismissed):
Since this is likely to be merged in before the end of the month, I wanted to add a general review/reminder to add a note about THM to the January newsletter now that `modules/doc/content/newsletter/2022_01.md` has made it to devel. Thanks!

loganharbour committed:
Remove .gitlab from old THM

loganharbour committed:
Remove .clang-format from old THM

loganharbour committed:
Remove .coverage from THM

loganharbour committed:
Port THM mailmap to mailmap

loganharbour committed:
Remove install-format-hook.sh from THM

permcody committed:
Restricting specific tests to replicated mesh mode

refs #19916

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/931448/) on 4e18938 wanted to post the following:

Warning: This PR changes repo size by 11.56 MiB.

loganharbour committed:
Make log_missing_test a warning for THM

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/932108/) on c5e1979 wanted to post the following:

Warning: This PR changes repo size by 11.56 MiB.

loganharbour committed:
Setup THM for SQA

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/932192/) on b550023 wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

loganharbour committed:
Use correct registrations for THM objects

loganharbour committed:
Remove FluidProperties in THM

loganharbour committed:
Remove FluidProperties/* action syntax in THM

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/932218/) on b550023 wanted to post the following:


## Framework coverage

Coverage did not change


## Modules coverage

Inconsistent report tags were found between the head and base reports.
This can happen when reports are missing from either the head or the base.

Inconsistent tags:
  module-thm
  module-thm-parallel
  module-thm-recover

### Combined

<table>
  <thead>
    <tr>
      <th></th>
      <th>50846a</th>
      <th colspan=4>#19916 b55002</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>37.96%</td>
      <td>74.14%</td>
      <td>+36.17%</td>
      <td>33.33%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>41</td>
      <td>43</td>
      <td>+2</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>67</td>
      <td>15</td>
      <td>-52</td>
      <td>6</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19916/coverage_diff/combined.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19916/coverage/combined)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/tensor_mechanics)
- [`thm`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/thm)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19916/coverage/xfem)

</details>

## Warnings

- `combined` new line coverage rate 33.33% is less than the suggested 90.0%



This comment will be updated on new commits.

loganharbour committed:
FluidProperties -> Modules/FluidProperties in THM tests

loganharbour committed:
FluidProperties -> Modules/FluidProperties in THM tutorials

loganharbour committed:
FluidProperties -> Modules/FluidProperties in THM templates

loganharbour committed:
FluidProperties -> Modules/FluidProperties in THM doco

loganharbour committed:
FluidProperties -> Modules/FluidProperties in THM python

loganharbour committed:
Switch moosedocs.py in THM to a symlink

loganharbour committed:
Reorder sqa report alphabetically

loganharbour committed:
Bring THM moosedocs python up to date

loganharbour committed:
FluidProperties -> Modules/FluidProperties in THM docs

loganharbour committed:
Add missing registrations in THM

loganharbour committed:
Get THM documentation working again

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/933910/) on 184c0c5 wanted to post the following:

Warning: This PR changes repo size by 11.56 MiB.

loganharbour committed:
Rename THM module to thermal_hydraulics

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/934093/) on a3eae8e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19916/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19916/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 18c73d2a68537676ad966764a8a2b319843a2a5e`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/934093/) on a3eae8e wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/934127/) on 12e46e5 wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/934202/) on c9f7196 wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

loganharbour committed:
Rename THM unit application

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/934811/) on 0c0a85b wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

loganharbour committed:
Rename thm -> thermal_hydraulics in tutorials

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/935283/) on 5e1ef72 wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

loganharbour committed:
Correct SQA link for thermal_hydraulics

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/935351/) on 0102b7a wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

loganharbour committed:
Fix sqa_thermal_hydraulics.yml

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/935536/) on 137ec58 wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

andrsd committed:
Moving away from alert2 to alert

lindsayad commented:
Any rough timeline on when we think this might be merged?

andrsd commented:
I resolved the doco issue yesterday, @loganharbour is still looking into some coverage issues. After that is resolved, we can merge. I think.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/937406/) on 11f53f3 wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

permcody committed:
Rearranging to fix dependency issues

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/937587/) on a884d63 wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

permcody committed:
Remove alert two from the TH module docs

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/937836/) on b4991a7 wanted to post the following:

Warning: This PR changes repo size by 11.57 MiB.

moosebuild commented:
Job [Python 3.8 Tests](https://civet.inl.gov/job/937840/) on b4991a7 : invalidated by @permcody

Error doesn&#39;&#39;t make any sense.

fdkong commented:
I would like to see this being merged. What can I help?

permcody commented:
Should we tell him @loganharbour? 😄

fdkong commented:
3K commits

loganharbour commented:
Something really weird is going on now that we've added this module. 

See the "Modules" recipe (which tests modules individually for coverage): https://civet.inl.gov/job/937847/ (6 hours) vs the current run on next https://civet.inl.gov/job/939598/ (1 hour). The difference is in the extraction of coverage. This went from taking < 20s per module to > 12m per module.

In specific, the final extraction (taking `chemical_reactions`) as an example:

```
BUILD_ROOT/moose/: lcov --gcov-tool gcov --extract /ramdisk/civet/build_0/moose_coverage/final.info \*/ramdisk/civet/build_0/moose/modules/chemical_reactions/src/\* --extract /ramdisk/civet/build_0/moose_coverage/final.info \*/ramdisk/civet/build_0/moose/modules/chemical_reactions/include/\* --output-file /ramdisk/civet/build_0/moose_coverage/coverage.info
Reading tracefile /ramdisk/civet/build_0/moose_coverage/final.info
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/include/auxkernels/KineticDisPreConcAux.h
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/include/auxkernels/KineticDisPreRateAux.h
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/include/utils/EquilibriumConstantFit.h
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/actions/AddCoupledEqSpeciesAction.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/actions/AddCoupledSolidKinSpeciesAction.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/actions/AddPrimarySpeciesAction.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/actions/AddSecondarySpeciesAction.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/AqueousEquilibriumRxnAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/EquilibriumConstantAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/KineticDisPreConcAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/KineticDisPreRateAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/PHAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/auxkernels/TotalConcentrationAux.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/base/ChemicalReactionsApp.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/bcs/ChemicalOutFlowBC.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/CoupledBEEquilibriumSub.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/CoupledBEKinetic.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/CoupledConvectionReactionSub.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/CoupledDiffusionReactionSub.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/DarcyFluxPressure.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/DesorptionFromMatrix.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/DesorptionToPorespace.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/PrimaryConvection.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/PrimaryDiffusion.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/kernels/PrimaryTimeDerivative.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/main.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/materials/LangmuirMaterial.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/materials/MollifiedLangmuirMaterial.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/postprocessors/TotalMineralVolumeFraction.C
Extracting /ramdisk/civet/build_0/moose/modules/chemical_reactions/src/utils/EquilibriumConstantFit.C
Extracted 30 files
Writing data to /ramdisk/civet/build_0/moose_coverage/coverage.info
Summary coverage rate:
  lines......: 93.3% (873 of 936 lines)
  functions..: 89.7% (96 of 107 functions)
  branches...: no data found
```

Is what takes many minutes. It will print out the summary and then hang. I have absolutely no clue what is going on here and I don't have the immediate time to debug it. Any takers? This should be able to be replicated on rod with:

```
moosebuild
module load mpich cmake lcov python vtk
git clone https://github.com/permcody/moose
cd moose
git checkout thm_merge_fixes_2
MOOSE_JOBS=32 ./scripts/update_and_rebuild_petsc.sh
METHODS=opt MOOSE_JOBS=32 ./scripts/update_and_rebuild_libmesh.sh
MOOSE_DIR=$(pwd)
cd framework
make -j 32
cd ../modules/chemical_reactions
MOOSE_HEADER_SYMLINKS=false make -j 32 coverage=true graceful=true
lcov --gcov-tool gcov --capture --initial --directory $MOOSE_DIR/modules/chemical_reactions --output-file $MOOSE_DIR/initialize.info
./run_tests -j 32
lcov --gcov-tool gcov --capture --ignore-errors gcov,source --directory $MOOSE_DIR/modules/chemical_reactions --output-file $MOOSE_DIR/covered.info
lcov --gcov-tool gcov --add-tracefile $MOOSE_DIR/initialize.info --add-tracefile $MOOSE_DIR/covered.info --output-file $MOOSE_DIR/final.info
lcov --gcov-tool gcov --extract $MOOSE_DIR/final.info $MOOSE_DIR/modules/chemical_reactions/src/* --extract $MOOSE_DIR/final.info $MOOSE_DIR/modules/chemical_reactions/include/* --output-file $MOOSE_DIR/coverage.info
```

The last command should hang. I typed all of the above without testing... it _should_ work

permcody committed:
Merge remote-tracking branch 'upstream/next' into thm_merge_fixes_2

permcody committed:
Moving THM documentation to more insane location so that those docs work properly

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/943531/) on 44578ce wanted to post the following:

Warning: This PR changes repo size by 11.63 MiB.
